### ISSUE TITLE
Algorithm does not recognize dictionary words in certain cases

### ISSUE THREAD

**Issue by fresheneesz** [dropbox/zxcvbn#274]
For example, `assboommoosemargaret` seems to work fine, showing 4 dictionary matches. However, if I add another dictionary word, it fails on the first word. For example, `assboommoosemargaretlook` does not recognize the first word and lumps the first two words into one brute force pattern. What's going on here?

**Comment by ghost** [comment]
Something similar happens with `therewaspaper` and `there.was.paper`. In the second `.was.` is considered to be a bruteforce pattern

**Comment by laughinghan** [comment]
I think this is behaving as intended.

By design, zxcvbn [heavily penalizes](https://github.com/dropbox/zxcvbn/blob/master/src/scoring.coffee#L52) additional components, i.e. it thinks an attacker is going to try a lot of guesses with 4 components before they start trying guesses with 5 components.

In this case, I think zxcvbn is estimating that an attacker will try a 4-component guess where one of the components is an up-to-7-letters "bruteforce" component (which will hit `assboom`), before they try a 5-dictionary-components attack. Similarly, I think zxcvbn is estimating that an attacker will try a 3-component guess where one of the components is an up-to-5-characters "bruteforce" (which will hit `.was.`), before they try a 5-component guess (zxcvbn considers each `.` a separate component).

Of course, this isn't tuned very precisely, in particular they estimate that a 7-letter bruteforce like `assboom` would only take 10<sup>7</sup> guesses, and that a 5-character bruteforce like `.was.` would only take 10<sup>5</sup> guesses, which are pretty low estimates, seeing as there are 26 letters, and that's not even counting symbols like `.`. [This is intentionally an underestimate](https://github.com/dropbox/zxcvbn/issues/17#issuecomment-150761549), since bruteforce is the fallback guess&mdash;what if there's a pattern that zxcvbn doesn't recognize, but that an attacker can take advantage of? An attacker almost certainly won't need 26<sup>7</sup> guesses to try `assboom`, and an attacker might well try separators like the `.`, a pattern that zxcvbn completely fails to recognize, but the bruteforce underestimate helps compensate for that.

**Comment by ghost** [comment]
My problem has been that I have found no good way of disallowing a user to use part of their email address.

email: john.doe@acme.com
dictionary: john, doe, acme
accepted password: john.doe@acme

This is a very predicable password and it should not be allowed. However, since the '.' and '@' causes an bruteforce pattern `.doe@` this password is accepted.

The case where this bit me was with a person with a longer surename, which gave enough "entropy" that it was put at score 3.

**Comment by laughinghan** [comment]
@albin-attentec Yeah, zxcvbn doesn't deal well with single-character separators, would be nice if they did. For your case, you could just, in addition to zxcvbn, search the string for their name and email I guess? (Excepting their name if it's really short, like "Jo")

**Comment by ghost** [comment]
yes, good point. I was just hoping that i would not have to.
It just seemed like a bug to me 🙂

**Comment by laughinghan** [comment]
@albin-attentec Have you seen the `user_inputs` feature? You can try passing the parts of the user's email address in as an array as the second argument: https://github.com/dropbox/zxcvbn#usage

(Based on some of the GitHub issues about it, it might not work that well, though)

**Comment by ghost** [comment]
Yes, that's the one that I am using.
I could try to slice the email in more ways, but it's probably best just to search for them manually instead.
Thank you for your help 🙂

**Comment by fresheneesz** [comment]
> zxcvbn heavily penalizes additional components

Sorry, I don't quite understand. Additional words should definitely make the password stronger, however, by treating two dictionary words as a pattern that must be brute forced does the *opposite* of that - vastly overestimating the strength of the password. I don't see how this could possibly be working as intended.

**Comment by laughinghan** [comment]
@fresheneesz I think you missed the part where the so-called "bruteforce" estimate actually estimates only ~3.3 bits of entropy per character, not something as naive as 8 bits or even 5 bits (which is the entropy of a random alphanumeric character). Specifically, even though there's obviously way more than 10 possible characters in a password (heck, just alphabetical letters is 26), in order to be conservative, they pretend there's just 10 possible characters (which works out to log<sub>2</sub>(10) ≈ 3.3 bits of entropy).

Maybe `bruteforce` is misnamed—maybe it should be called it `markov`, because the choice of 10 was specifically [chosen with a markov attack in mind, not a true naive bruteforce](https://github.com/dropbox/zxcvbn/issues/17#issuecomment-150761549).

(I think it's pretty easy to augment a markov attack to cover the entire search space, turning it into basically a search-order-optimized bruteforce. So I think it's safe to assume no attacker would ever naively bruteforce a long password, and therefore I think it makes sense for the fallback estimate should be markov entropy, not any kind of bruteforce entropy.)

**Comment by LHSnow** [comment]
I think I ran into a similar case, as the following user input still scores 3. I guess I am assuming direct matches from user input would penalise the score more than it does.

```javascript
zxcvbn('john_doned1957', ['john','doned','1957'])
```
